tests/network/mod.rs

```diff
@@ -601,21 +608,59 @@ where
     D: DistAlgorithm<Output = Batch<C, NodeId>, NodeId = NodeId>,
     C: Contribution + Clone,
 {
-    /// Verifies that all nodes' outputs agree.
+    /// Verifies that all nodes' outputs agree, given a correct "full" node that output all
+    /// batches with no gaps.
+    ///
+    /// The output of the full node is used to derive in expected output of other nodes in every
+    /// epoch. After that the check ensures that correct nodes output the same batches in epochs
+    /// when those nodes were participants (either validators or candidates).
     #[allow(unused)] // Not used in all tests.
-    pub fn verify_batches(&self) {
-        let expected = self.nodes[&NodeId(0)].outputs().to_vec();
-        assert!(!expected.is_empty());
-        let pub_eq = |(b0, b1): (&Batch<C, _>, &Batch<C, _>)| b0.public_eq(b1);
-        for node in self.nodes.values() {
-            assert_eq!(expected.len(), node.outputs().len());
+    pub fn verify_batches<E>(&self, full_node: &TestNode<D>)
+    where
+        Batch<C, NodeId>: SenderQueueableOutput<NodeId, E>,
+    {
+        // Participants of epoch 0 are all validators in the test network.
+        let mut participants: BTreeSet<NodeId> = self
+            .nodes
+            .keys()
+            .cloned()
+            .chain(self.adv_nodes.keys().cloned())
+            .collect();
+        let mut expected: BTreeMap<NodeId, Vec<_>> = BTreeMap::new();
+        for batch in &full_node.outputs {
+            for id in &participants {
+                expected.entry(id.clone()).or_default().push(batch);
+            }
+            if let Some(new_participants) = batch.participant_change() {
+                participants = new_participants;
+            }
+        }
+        for (id, node) in self.nodes.iter().filter(|(&id, _)| id != full_node.id) {
+            let actual_epochs: BTreeSet<_> =
+                node.outputs.iter().map(|batch| batch.epoch()).collect();
+            let expected_epochs: BTreeSet<_> =
+                expected[id].iter().map(|batch| batch.epoch()).collect();
+            assert_eq!(
+                expected_epochs, actual_epochs,
+                "Output epochs of {:?} don't match the expectation.",
+                id
+            );
             assert!(
-                expected.iter().zip(node.outputs()).all(pub_eq),
-                "Outputs of nodes 0 and {} differ: {:?} != {:?}",
-                node.instance().our_id().0,
-                expected,
-                node.outputs()
+                node.outputs
+                    .iter()
+                    .zip(expected.get(id).expect("node is not expected"))
+                    .all(|(a, b)| a.public_eq(b)),
+                "Outputs of {:?} don't match the expectation",
+                id
             );
         }
+        assert!(
```

Comment:
Since `zip` doesn't fail if the lengths differ, we should assert that `self.observer.outputs.len()` equals `full_node.outputs.len()`.